### PR TITLE
sql/mssql: initial mssql driver

### PR DESCRIPTION
**Pull request by giautm**

### CODE REVIEW

File: internal/ci/main.go [hunk 1]
```diff
@@ -161,6 +161,26 @@ var (
...
+			Regex:   "SQLServer",
+			Ports:   []string{"1433:1433"},
+			Env: []string{
+				"ACCEPT_EULA: Y",
+				"MSSQL_SA_PASSWORD: Passw0rd!995",
+			},
+		},
+		{
+			Version: "azure-sql-edge",
+			Image:   "mcr.microsoft.com/azure-sql-edge:latest",
```
Comment (masseelch) [inline internal/ci/main.go:176]: Shouldn't we pinpoint a specific version instead of latest?

File: schemahcl/types.go [hunk 2]
```diff
@@ -343,6 +343,24 @@ func SizeTypeAttr(required bool) *TypeAttr {
...
+		Required: false,
+	}
+}
+
+// PrecisionTypeAttr returns a TypeAttr for a precision attribute.
+func PrecisionTypeAttr() *TypeAttr {
+	return &TypeAttr{
+		Name:     "precision",
+		Kind:     reflect.Int,
+		Required: false,
```
Comment (masseelch) [inline schemahcl/types.go:360]:
```suggestion
```

File: schemahcl/types.go [hunk 3]
```diff
@@ -343,6 +343,24 @@ func SizeTypeAttr(required bool) *TypeAttr {
 	}
 }
 
+// ScaleTypeAttr returns a TypeAttr for a scale attribute.
+func ScaleTypeAttr() *TypeAttr {
+	return &TypeAttr{
+		Name:     "scale",
+		Kind:     reflect.Int,
+		Required: false,
```
Comment (masseelch) [inline schemahcl/types.go:351]:
```suggestion
```

File: sql/mssql/convert.go [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+)
+
+type (
+	// BitType defines a bit type.
+	// https://learn.microsoft.com/en-us/sql/t-sql/data-types/bit-transact-sql
+	BitType struct {
+		schema.Type
+		T string
+	}
+	// A HierarchyIDType defines a hierarchyid type.
```
Comment (masseelch) [inline sql/mssql/convert.go:22]:
```suggestion
	// HierarchyIDType defines a hierarchyid type.
```

File: sql/mssql/convert.go [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+	// https://learn.microsoft.com/en-us/sql/t-sql/data-types/hierarchyid-data-type-method-reference
+	HierarchyIDType struct {
+		schema.Type
+		T string
+	}
+	// A MoneyType defines a money type.
```
Comment (masseelch) [inline sql/mssql/convert.go:28]:
```suggestion
	// MoneyType defines a money type.
```

File: sql/mssql/convert.go [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+	UniqueIdentifierType struct {
+		schema.Type
+		T string
+	}
+	// UserDefinedType defines a user-defined type attribute.
+	UserDefinedType struct {
+		schema.Type
+		T string
+	}
+	// A XMLType defines an XML type.
```
Comment (masseelch) [inline sql/mssql/convert.go:45]:
```suggestion
	// XMLType defines an XML type.
```

File: sql/mssql/convert.go [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+		schema.Type
+		T string
+	}
+	// A MoneyType defines a money type.
+	// https://learn.microsoft.com/en-us/sql/t-sql/data-types/money-and-smallmoney-transact-sql
+	MoneyType struct {
+		schema.Type
+		T string
+	}
+	// A UniqueIdentifierType defines a uniqueidentifier type.
```
Comment (masseelch) [inline sql/mssql/convert.go:34]:
```suggestion
	// UniqueIdentifierType defines a uniqueidentifier type.
```

File: schemahcl/types.go [hunk 8]
```diff
@@ -343,6 +343,24 @@ func SizeTypeAttr(required bool) *TypeAttr {
 	}
 }
 
+// ScaleTypeAttr returns a TypeAttr for a scale attribute.
+func ScaleTypeAttr() *TypeAttr {
```
Comment (a8m) [inline schemahcl/types.go:347]: This is a driver-specific attribute (right?). Let's keep it in mssql.

File: schemahcl/types.go [hunk 9]
```diff
@@ -343,6 +343,24 @@ func SizeTypeAttr(required bool) *TypeAttr {
...
+	return &TypeAttr{
+		Name:     "scale",
+		Kind:     reflect.Int,
+		Required: false,
+	}
+}
+
+// PrecisionTypeAttr returns a TypeAttr for a precision attribute.
```
Comment (a8m) [inline schemahcl/types.go:355]: ditto

File: sql/internal/sqlx/diff.go [hunk 10]
```diff
@@ -476,6 +476,28 @@ func Has(elements, target any) bool {
...
+// HasCharset reports if the attribute contains the "charset" attribute,
+// and it needs to be defined explicitly on the schema. This is true, in
+// case the element charset is different from its parent charset.
+func HasCharset(attr []schema.Attr, parent []schema.Attr) (string, bool) {
+	var c, p schema.Charset
+	if Has(attr, &c) && (parent == nil || Has(parent, &p) && c.V != p.V) {
+		return c.V, true
+	}
+	return "", false
+}
```
Comment (a8m) [inline sql/internal/sqlx/diff.go:488]: Remove as it's not used.

File: schemahcl/types.go (hunk 8 above)
Comment (giautm) [reply schemahcl/types.go:347]: I saw it repeated in mysql, pg too

File: sql/internal/sqlx/diff.go [hunk 11]
```diff
@@ -476,6 +476,28 @@ func Has(elements, target any) bool {
...
+
+// HasCollate reports if the attribute contains the "collation"/"collate" attribute,
+// and it needs to be defined explicitly on the schema. This is true, in
+// case the element collation is different from its parent collation.
+func HasCollate(attr []schema.Attr, parent []schema.Attr) (string, bool) {
```
Comment (a8m) [inline sql/internal/sqlx/diff.go:493]: If it's copied from mysql, let's do it in a separate PR and reuse it there as well.

Comment (giautm) [reply sql/internal/sqlx/diff.go:493]: Sure, will move to another pr for small work on this

File: sql/mssql/convert.go [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"ariga.io/atlas/sql/schema"
+)
+
+type (
```
Comment (a8m) [inline sql/mssql/convert.go:15]: Keep types in inspect.go file. Like mysql, postgres and sqlite.

File: schemahcl/types.go (hunk 8 above)
Comment (a8m) [reply schemahcl/types.go:347]: kk, same as "HasCollate" - let's do it in a separate PR and move it to a shared place. To keep driver PRs, mssql specific.

File: sql/mssql/convert.go [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+		c.typ = s
+	}
+	return c, nil
+}
+
+func columnType(c *columnDesc) (schema.Type, error) {
+	size, precision, scale := int(c.size), int(c.precision), int(c.scale)
+	var typ schema.Type
+	switch t := c.typ; strings.ToLower(t) {
+	// numeric types
```
Comment (a8m) [inline sql/mssql/convert.go:264]:
```suggestion
```

File: sql/mssql/inspect.go [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+	SELECT
+	[schema_name] = [name]
+	FROM [sys].[schemas]
+	WHERE [name] NOT IN (
+		'db_accessadmin', 'db_backupoperator', 'db_datareader',
+		'db_datawriter', 'db_ddladmin', 'db_denydatareader',
+		'db_denydatawriter', 'db_owner', 'db_securityadmin',
+		'guest', 'information_schema', 'sys'
+	)
+	ORDER BY [schema_id]`
```
Comment (a8m) [inline sql/mssql/inspect.go:281]: Not schema name?

File: sql/mssql/inspect.go [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+	FROM [sys].[tables] as [t1]
+	LEFT JOIN [sys].[extended_properties] [td]
+	ON [td].[major_id] = [t1].[object_id]
+	AND [td].[minor_id] = 0
+	AND [td].[name] = N'MS_Description'
+	WHERE SCHEMA_NAME([t1].[schema_id]) IN (%s)
+	AND [t1].[type] = 'U'
+	AND [t1].[is_ms_shipped] = 0
+	ORDER BY [table_schema], [table_name]`
+
```
Comment (a8m) [inline sql/mssql/inspect.go:309]:
```suggestion
SELECT
	[table_schema] = SCHEMA_NAME([t1].[schema_id]),
	[table_name] = [t1].[name],
	[comment] = [td].[value]
FROM
	[sys].[tables] as [t1]
	LEFT JOIN [sys].[extended_properties] [td]
	ON [td].[major_id] = [t1].[object_id]
	AND [td].[minor_id] = 0
	AND [td].[name] = N'MS_Description'
WHERE
	SCHEMA_NAME([t1].[schema_id]) IN (%s)
	AND [t1].[type] = 'U'
	AND [t1].[is_ms_shipped] = 0
ORDER BY
	[table_schema], [table_name]`

```

Comment (a8m) [reply sql/mssql/inspect.go:309]: Prettify the others as well :)

File: sql/mssql/inspect.go (hunk 14 above)
Comment (giautm) [inline sql/mssql/inspect.go:281]:
```suggestion
	ORDER BY [schema_name]`
```
